Snapshot hidden ref

kwryankrattiger opened the pull request:
Spack has some problems around snapshots

1. This script is pointing at the wrong repo after repo split
    - Github: `spack/spack` -> `spack/spack-packages`
    - Gitlab: 2 -> 57
2. The current method of pinning snapshots uses tags, which creates extra noise around the GH releases feature
    - Tags imply (or should imply) immutability
3. Spack Core Gitlab CI is unstable as it points at the head of packages develop, which may or may not be passing in packages
    - Needs to either manually update packages ref
    - Need a moving ref. (shouldn't be a tag)

Spack snapshots have historically been pushed as tags which can be noisy in Github since tags and releases are tied together. Alternatives to tag refs are either branches or hidden refs. While most people are more familiar with branches, a hidden ref will not inundate the branch search. For that reason, we decided to switch to hidden refs.

The Spack core repo can now point at `snapshots/develop-latest` for packages, which is guaranteed to have at least passed the packages CI check. The downside to this is extended periods of broken packages develop will end up delaying updates in core, this is worth it to ensure gitlab checks can be made required in core.

kwryankrattiger committed:
Remove unused requirements from snapshot-release-tags

kwryankrattiger committed:
Make snapshots hidden refs instead of tags

This will remove snapshot tags.

kwryankrattiger committed:
Update projects to point at packages

Only running snapshot for packages commits

kwryankrattiger committed:
Don't update if already the latest

kwryankrattiger committed:
Run the snapshot more often.

Snapshots are just index views of the cache so we can make them more
often.

kwryankrattiger committed:
And it seems to work

kwryankrattiger committed:
Every third day at 1am

kwryankrattiger committed:
Bump image versions

zackgalbreath commented on images/gh-gl-sync/SpackCIBridge.py:
```diff
@@ -767,4 +795,4 @@ def sync(self):
                            required_label=args.required_label)
     bridge.setup_ssh(ssh_key_base64)
-    bridge.sync()
+    bridge.sync(hidden_refs=args.hidde_ref)
```
```suggestion
    bridge.sync(hidden_refs=args.hidden_ref)
```

zackgalbreath commented on images/snapshot-release-tags/snapshot_release_tags.py:
```diff
@@ -36,27 +40,39 @@
+            print("Latest ref is already latest snapshot")
+            sys.exit(0)
+    except GithubException:
+        # Failure to get the latest_ref means it doesn't exist and needs to be recreated.
```
```suggestion
    except GithubException as e:
        if e.status != 404:
            raise
        # 404 means the ref doesn't exist yet and needs to be created.
```

zackgalbreath commented on images/snapshot-release-tags/snapshot_release_tags.py:
```diff
@@ -36,27 +40,39 @@
+    except GithubException:
+        # Failure to get the latest_ref means it doesn't exist and needs to be recreated.
+        latest_ref = None
+        pass
```
```suggestion
```

zackgalbreath commented on images/snapshot-release-tags/snapshot_release_tags.py:
```diff
@@ -36,27 +40,39 @@
+    if latest_ref is not None:
+        latest_ref.edit(sha, force=True)
+    else:
+        py_gh_repo.create_git_ref(
```
It might make sense to catch "ref already exists" here (rather than letting it propagate to sentry) but reading [the docs](https://docs.github.com/en/rest/git/refs?apiVersion=2026-03-10#create-a-reference--status-codes) I'm not sure if that would be 409 or 422 (or if we would even be able to cleanly distinguish this type of error from others). WDYT?

kwryankrattiger replied:
For this one we are updating the `snapshots/develop-latest` which, in most cases, we expect to exist. The issue is you need a different API call to update the ref vs push a new one.

I did miss above, where I am blindly updating the date ref, that we may snapshot multiple refs on the same day now. That I can fix by only running the cron once a day, but probably a good idea to make it a little more robust than that.

kwryankrattiger committed:
Undo changes to sync script

These changes are no longer needed for snapshots, copy-only job is now
run by a cron.

kwryankrattiger committed:
Do a better job handling exceptions when updating refs

zackgalbreath commented on images/snapshot-release-tags/snapshot_release_tags.py:
```diff
@@ -36,27 +74,15 @@
+    # If the latest develop snapshot is updated, also push a new date snapshot
+    if create_or_update_ref(repo, "snapshots/develop-latest", sha):
+        date_str = datetime.now(timezone.utc).strftime("%Y-%m-%d")
+        ref_name = f"develop-{date_str}"
```
```suggestion
        ref_name = f"snapshots/develop-{date_str}"
```
